1-Draft/RFCNNNN-Automatic-Resource-Management.md

```diff
@@ -0,0 +1,197 @@
+---
+RFC: RFC<four digit unique incrementing number assigned by Committee, this shall be left blank by the author>
+Author: Ilya Sazonov
+Status: Draft
+SupercededBy: <link to another RFC>
+Version: 1.0
+Area: Engine
+Comments Due: 7/1/2020
+Plan to implement: Maybe
+---
+
+# Automatic resource management
+
+## Rationale
+
+Currently PowerShell has limited capabilities compared with C# to excecute a mandatory code.
+
+PowerShell users can benefit from `finally` block of `try - catch` statement. It is very limited experience because resources in PowerShell can be distributed across scopes, cmdlets, pipelines and runspaces.
+
+If an user created a temporary file one should worry about deleting it, while PowerShell could register a mandatory code to delete this file and execute it automatically at the right time.
+
+PowerShell makes many smart things for users and could offer smart resource management.
+
+## Motivation
+
+As a script writer, I can use some well-known API-s which grabs resources
+and I want PowerShell understands this and releases the resources automatically at the right time
+so that avoids resorce leaks.
+
+As an advanced module and script writer, I can use some API-s which grabs resources
+and I want to register a code PowerShell must run to release the resources automatically at the right time so that avoids resorce leaks.
+
+As an advanced module and script writer, I want to make sure that a mandatory (cleanup, logging) code is always executed even if a pipeline / script / cmdlet is unexpectedly interrupted.
```

Comment:
@vexx32 All your comment comes from  #207 RFC position. It does not add value in the discussion. Current RFC is about another _concept_. If you want to discuss the concept you should forget that RFC and freely play with the concept.

> It's also significantly less versatile

No, developer can register any complex code like in C#. Rather, we should even limit it so that this code really does the cleanup and nothing more.

Reply:
Understood. I don't know how you'd go about limiting that in any appreciable manner, though.